contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md

```diff
@@ -0,0 +1,130 @@
+
+This code should work on a regular CPU (in a docker container), which will allow you to test out the code locally without
+any additional setup. This specific approach this code uses is [LoRA](https://arxiv.org/abs/2106.09685) (low-rank adaptation of large language models), which
+means that only a subset of the LLM's parameters are tweaked and prevents over-fitting.
+
+## What is fine-tuning?
+Fine-tuning is when a pre-trained model, in this context a foundational model, is customized using additional data to
+adjust its responses for a specific task. This is a good way to adjust an off-the-shelf, i.e. pretrained model, to provide
+more responses that are more contextually relevant to your use case.
+
+## FLAN LLM
+This example is based on using [Google's Fine-tuned LAnguage Net (FLAN) models hosted on HuggingFace](https://huggingface.co/docs/transformers/model_doc/flan-t5).
+The larger the model, the longer it will take to fine-tune, and the more memory you'll need for it. The code
+here by default (which you can easily change) is set up to run on docker using the smallest FLAN model.
+
+## What type of functionality is in this module?
+
+The module uses libraries such as numpy, pandas, plotly, torch, sklearn, peft, evaluate, datasets, and transformers.
+
+It shows a basic process of:
+
+a. Loading data and tokenizing it and setting up some tokenization parameters.
+
+b. Splitting data into training, validation, and inference sets.
+
+c. Fine-tuning the model using LoRA.
+
+d. Evaluating the fine-tuned model using the [rouge metric](https://en.wikipedia.org/wiki/ROUGE_(metric)).
+
+You should be able to read the module top to bottom which corresponds roughly to the order of execution.
+
+## How might I use this module?
+To use this module you'll need to do the following:
+
+- Data. The data set should be list of JSON objects, where each entry in the list is an object that has the "question" and
+and also the "answer" to that question. You will provide the name of the keys for these fields as input to run the code.
+e.g. you should be able to do `json.load(f)` and it would return a list of dictionaries, e.g. something like this:
+
+```python
+[
+    {
+        "question": "What is the meaning of life?",
+        "reply": "42"
+    },
+    {
+        "question": "What is Hamilton?",
+        "reply": "..."
+    },
+    ...
+]
+```
+
+You would then pass in as _inputs_ to execution `"data_path"=PATH_TO_THIS_FILE` as well as `"input_text_key"="question"` and `"output_text_key"="reply"`.
+- Instantiate the driver. Use `{"start": "base"}` as configuration to run with to use a raw base LLM to finetune.
+- Pick your LLM. `model_id_tokenizer="google/mt5-small"` is the default, but you can change it to any of the models
+that the transformers library supports for `AutoModelForSeq2SeqLM` models.
+- Run the code.
+
+Because there's no configuration that changes the shape of the DAG, you can run the code like this:
```

Comment:
Clarify "Becuase there's no configuration that changes the shape of the DAG" (remove)